SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp

```diff
@@ -0,0 +1,289 @@
+	bool GetNewestLogDir = IFileManager::Get().IterateDirectoryStat(*LocalDeploymentLogsDir, [&NewestLogDir, &NewestLogDirTime](const TCHAR* FileName, const FFileStatData& FileStats) {
+
+		if (FileStats.bIsDirectory)
+		{
+			if (FileStats.CreationTime > NewestLogDirTime)
+			{
+				NewestLogDir = FString(FileName);
+				NewestLogDirTime = FileStats.CreationTime;
+			}
+		}
+
+		return true;
+	});
+
+	if (GetNewestLogDir)
+	{
+		StartPollingLogFile(FPaths::Combine(NewestLogDir, LaunchLogFilename));
+	}
+}
+
+SSpatialOutputLog::~SSpatialOutputLog()
+{
+	CloseLogReader();
+
+	ShutdownLogDirectoryWatcher(LocalDeploymentLogsDir);
+
+	FCoreDelegates::OnHandleSystemError.RemoveAll(this);
+}
+
+void SSpatialOutputLog::OnCrash()
+{
+	CloseLogReader();
+
+	ShutdownLogDirectoryWatcher(LocalDeploymentLogsDir);
+}
+
+void SSpatialOutputLog::StartUpLogDirectoryWatcher(FString LogDirectory)
+{
+	AsyncTask(ENamedThreads::GameThread, [this, LogDirectory]
+	{
+		FDirectoryWatcherModule& DirectoryWatcherModule = FModuleManager::LoadModuleChecked<FDirectoryWatcherModule>(TEXT("DirectoryWatcher"));
+		if (IDirectoryWatcher * DirectoryWatcher = DirectoryWatcherModule.Get())
+		{
+			// Watch the log directory for changes.
+			if (FPaths::DirectoryExists(LogDirectory))
+			{
+				LogDirectoryChangedDelegate = IDirectoryWatcher::FDirectoryChanged::CreateRaw(this, &SSpatialOutputLog::OnLogDirectoryChanged);
+				DirectoryWatcher->RegisterDirectoryChangedCallback_Handle(LogDirectory, LogDirectoryChangedDelegate, LogDirectoryChangedDelegateHandle, IDirectoryWatcher::WatchOptions::IncludeDirectoryChanges | IDirectoryWatcher::WatchOptions::IgnoreChangesInSubtree);
+			}
+			else
+			{
+				UE_LOG(LogSpatialOutputLog, Error, TEXT("Local deployment log directory does not exist!"));
+			}
+		}
+	});
+}
+
+void SSpatialOutputLog::OnLogDirectoryChanged(const TArray<FFileChangeData>& FileChanges)
+{
+	// If this is a new folder creation then switch to watching the log files in that new log folder.
+	for (FFileChangeData FileChange : FileChanges)
+	{
+		if (FileChange.Action == FFileChangeData::FCA_Added)
+		{
+			// Now we can start reading the new log file in the new log folder.
+			StartPollingLogFile(FPaths::Combine(FileChange.Filename, LaunchLogFilename));
+		}
+	}
+}
+
+void SSpatialOutputLog::ShutdownLogDirectoryWatcher(FString LogDirectory)
+{
+	AsyncTask(ENamedThreads::GameThread, [this, LogDirectory]
+	{
+		FDirectoryWatcherModule& DirectoryWatcherModule = FModuleManager::LoadModuleChecked<FDirectoryWatcherModule>(TEXT("DirectoryWatcher"));
+		if (IDirectoryWatcher * DirectoryWatcher = DirectoryWatcherModule.Get())
+		{
+			DirectoryWatcher->UnregisterDirectoryChangedCallback_Handle(LogDirectory, LogDirectoryChangedDelegateHandle);
+		}
+	});
+}
+
+void SSpatialOutputLog::CloseLogReader()
+{
+	if (GEditor != nullptr)
+	{
+		// Delete the old timer if one exists.
+		GEditor->GetTimerManager()->ClearTimer(PollTimer);
```

Comment:
We should also wait for this task to finish.

Reply:
Looking at this I think this could cause things to break. If you are waiting for the next poll then this is fine. If not then you might call close on the log reader as it is reading. I think you could even clear the timer just before it sets a new timer. Then if you closed the reader because there should be a new one you'd end up calling it too often.
This should set a flag saying that the logger needs to stop, then wait for it to stop, then clear the timer.